README.md
# myte - A Tiny Chess Engine

Myte is a tiny C# chess bot created for the [Chess Coding Challenge](https://github.com/SebLague/Chess-Challenge) by [Sebastian Lague](https://www.youtube.com/@SebastianLague).

- [CHANGELOG.md](CHANGELOG.md) contains a history of feature changes and test results

## Features

- Engine
  - Time Control Management
- Search
  - Negamax Alpha-Beta search
  - Principal Variation Search
  - Quiescence Search
  - Iterative Deepening
  - Transposition Table
  - Check Extensions
  - Move Ordering
    - Transposition Table Moves
    - MVV-LVA
    - Killer Moves Heuristic
    - Countermove Heuristic
  - Null Move Pruning
  - Static Null Move Pruning
- Evaluation
  - Tapered Evaluation
  - Material Evaluation
  - Piece-Square Tables
  - Bishop Pairs
- Tuning
  - Texel Tuning with gradient descent
 
## Makefile Targets
```
build           Build all executables
build-gui       Build the regular chess GUI
build-uci       Build the UCI program interface

run             Run the regular chess GUI
uci             Run the bot in UCI mode
uci-baseline    Run the baseline bot in UCI mode

baseline        Copy the source code of the current bot to the baseline bot
```


docs/CHANGELOG.md
# Changelog

## v1.0.0
```
Use transposition table in quiescence search.

Time Control:   inf/10+0.0
Brain Capacity: 1019/1024 (99.51% full) [-3 / -0.29%]

Score of Myte 1.0.0 vs Myte 1.0.0-baseline: 359 - 275 - 292  [0.545] 926
...      Myte 1.0.0 playing White: 176 - 146 - 141  [0.532] 463
...      Myte 1.0.0 playing Black: 183 - 129 - 151  [0.558] 463
...      White vs Black: 305 - 329 - 292  [0.487] 926
Elo difference: 31.6 +/- 18.5, LOS: 100.0 %, DrawRatio: 31.5 %
SPRT: llr 2.98 (101.4%), lbound -2.94, ubound 2.94 - H1 was accepted
Elo Gain Bounds: [13.1, 50.1]
```

## v0.16.0 Endgame Material Values + Texel Tuning
```
Add endgame material values to taper and tune all values.

Texel tuning on 139,891 positions
    MSE Before: 0.377726231
    MSE After:  0.376393310 (-0.0013329206178178121)

Time Control:   inf/10+0.0
Brain Capacity: 1022/1024 (99.80% full) [+31 / +3.03%]

Score of Myte 0.16.0 vs Myte 0.16.0-baseline: 448 - 355 - 250  [0.544] 1053
...      Myte 0.16.0 playing White: 237 - 164 - 126  [0.569] 527
...      Myte 0.16.0 playing Black: 211 - 191 - 124  [0.519] 526
...      White vs Black: 428 - 375 - 250  [0.525] 1053
Elo difference: 30.8 +/- 18.4, LOS: 99.9 %, DrawRatio: 23.7 %
SPRT: llr 2.95 (100.1%), lbound -2.94, ubound 2.94 - H1 was accepted
Elo Gain Bounds: [12.4, 49.2]
```

## v0.15.1 Simplify Transposition Table
```
(Regression check) Reduce token usage in transposition table by collapsing to single case

Time Control:   inf/10+0.0
Brain Capacity: 991/1024 (96.78% full) [-24 / -2.34%]


tools/encode.py
'''
This script encodes arbitrary score tables into compressed bitmaps that can be expanded at runtime during initialization.

Observation: For any given table, the range between the min and max values is generally less than 128 (2^7). 

We can encode the table as a series of n bitmaps, where each bitmap represents a bit value (2^i for i=0..n) and each bit 
in the bitmap represents a square. Then to recover the original value for square i, we can look at the i'th bit of each 
bitmap and add up that bitmap's 2^n value if it's set.

Observation: Although the table ranges are <128, they don't all span the same 128 integer range.

This is a problem since it means we can't just sum of the bitmap values and add the result to the absolute minimum among
all tables, as any larger values outside of that minimum + 128 wouldn't be representable. To remedy this we'll use only
6 bitmaps instead (giving 2^6 = 64 encoding range) and encode an additional score table that is used to store the
minimum values for each table to use as their "zeroes".
'''

from dataclasses import dataclass

@dataclass
class ScoreTable:
    name: str
    table: [int]
    compressed: [int]
# piece move
# piece capture
# en passant
# castling (qk)
# promotion
# combos (capture + promotion)

phases = ['midgame', 'endgame']
pieces = ['pawn', 'knight', 'bishop', 'rook', 'queen', 'king']

# Done, improved by 0.0013329206178178121 to 0.37639331011680077

# [100,315,324,410,977,0, 124,200,229,429,687,0],
# 0, 100,338,344,439,1044,0, 126,216,242,460,735,0
# 0, 100,335,343,436,1037,0, 125,213,242,457,730,0

tables = [
# [0,0,0,0,0,0,0,0,43,42,41,46,29,24,24,30,17,11,17,28,29,19,19,-3,-32,-16,-16,-6,-3,-14,-12,-15,-43,-31,-24,-10,-15,-15,-12,-33,-43,-31,-35,-29,-26,-19,0,-31,-47,-38,-36,-68,-57,-1,8,-35,0,0,0,0,0,0,0,0],
# [-8,-4,-1,0,3,0,0,-8,-16,-9,8,7,7,8,-4,-11,-8,8,18,15,19,27,10,6,-3,11,9,34,34,58,6,0,-24,1,18,11,30,23,10,-10,-35,-15,0,5,11,6,16,-22,-6,-22,-16,-4,-7,4,-4,-5,-5,-37,-14,-34,-14,-12,-27,0],
# [0,0,0,-4,0,0,0,0,-16,0,-10,-4,0,5,0,-7,-5,8,-1,5,6,15,11,20,-6,1,3,28,16,19,3,-4,-8,0,10,34,23,8,-4,-9,-1,3,8,0,0,-2,4,3,1,-6,0,-10,0,10,11,-15,-13,-12,-19,-19,-16,-26,-6,-3],
# [3,2,7,12,3,5,10,10,-6,-4,21,24,20,18,8,10,-15,-4,6,11,13,17,16,5,-18,-7,-5,7,9,9,-4,-8,-34,-30,-9,-4,5,4,10,-14,-31,-21,-6,-9,-2,7,6,-6,-49,-13,-14,-6,-1,15,-10,-32,-3,-5,-9,7,12,14,-13,4],
# [-10,3,4,9,16,4,3,2,-12,-16,9,18,9,8,6,16,-23,-7,1,19,24,32,25,31,-12,-8,3,10,21,23,16,16,-13,-9,-9,3,11,12,13,9,-4,-3,-3,-4,4,9,18,2,-16,-16,2,3,2,2,-22,-14,-12,-15,-19,0,-11,-27,-7,-11],
# [-4,-4,0,-4,0,0,5,4,0,6,4,4,1,4,2,4,4,8,9,5,4,3,8,1,2,8,7,10,5,6,6,-6,-1,4,0,3,-1,1,0,-8,-2,-2,-3,-6,-20,-2,-2,-9,-4,-3,3,-19,-39,-30,16,15,-18,50,6,-22,-7,-56,33,20],
# [0,0,0,0,0,0,0,0,90,104,93,86,68,85,74,91,34,57,44,29,32,19,20,20,-4,-7,-23,-34,-40,-29,-6,-23,-18,-34,-34,-45,-45,-39,-24,-33,-30,-29,-41,-33,-36,-35,-33,-36,-20,-20,-27,-28,-19,-34,-25,-35,0,0,0,0,0,0,0,0],
# [-8,-5,-5,0,0,-5,-5,-12,-13,-2,8,5,9,-5,-11,-10,-15,2,26,20,15,22,3,-6,-7,5,22,31,0,18,7,-7,-7,-13,21,19,15,9,-5,-5,-11,-11,0,10,5,-4,-5,-13,-9,-16,-21,-9,-1,3,-1,-9,-13,-16,-20,-22,-10,-15,-11,-5],
# [-2,-6,0,-8,3,-1,-1,-1,-6,3,0,-1,0,0,3,3,0,2,-2,0,0,20,10,-2,-4,9,8,11,8,7,-3,8,-3,2,16,8,3,2,-8,-10,-2,-7,14,20,16,-2,-11,-3,-10,-5,-1,4,-14,-4,-20,-15,-11,-8,-24,-20,-8,-19,-6,-5],
# [6,15,11,12,11,12,8,10,15,19,15,18,18,16,13,14,14,15,14,12,12,17,13,2,11,11,15,11,10,11,2,4,3,4,3,3,-3,-1,-10,-24,2,-13,-14,-14,-15,-11,-16,-25,-25,-12,-15,-14,-15,-10,-24,-27,-18,-16,-10,-17,-21,-22,-25,-38],
# [-21,-4,2,13,13,-4,-11,-9,-6,-4,9,15,13,4,4,2,-14,1,6,19,22,16,7,10,-5,2,5,13,22,21,11,11,-7,-3,0,17,1,8,2,0,-4,-7,-7,-7,-7,-8,0,-3,-7,-14,-14,-24,-15,-16,-18,-12,-10,-19,-22,-35,-23,-32,-23,-33],
# [-34,-28,-8,-21,-12,5,19,16,-7,14,15,7,5,24,14,10,4,6,23,5,6,8,21,0,2,16,22,17,13,24,7,-12,-10,0,21,11,19,18,0,-21,-11,-6,10,6,21,12,-3,-16,-18,-12,8,7,11,12,-7,-21,-37,-41,-16,-20,-19,-4,-32,-40],

# [0,0,0,0,0,0,0,0,43,42,41,46,29,24,24,30,17,11,17,28,29,19,19,-3,-32,-16,-16,-7,-3,-14,-12,-15,-43,-31,-24,-11,-15,-15,-12,-33,-43,-31,-35,-29,-26,-18,0,-31,-47,-39,-36,-68,-57,-2,7,-35,0,0,0,0,0,0,0,0],
# [-8,-4,-1,0,3,0,0,-8,-16,-9,8,7,7,8,-4,-11,-8,8,18,15,19,27,10,6,-3,11,9,34,34,58,6,0,-24,1,18,11,30,23,10,-9,-35,-16,0,5,11,6,16,-22,-6,-22,-16,-4,-7,4,-4,-5,-5,-37,-14,-34,-14,-12,-27,-1],
# [0,0,0,-4,0,0,0,0,-16,0,-10,-4,0,5,0,-7,-5,8,-1,5,6,15,11,20,-6,1,3,28,16,19,3,-4,-8,0,10,34,23,8,-4,-9,-1,3,8,0,0,-2,4,3,1,-6,0,-10,0,10,11,-15,-13,-12,-18,-19,-16,-26,-6,-3],
# [3,2,7,12,3,5,10,10,-6,-4,21,24,20,18,8,10,-15,-4,6,11,13,17,16,5,-18,-7,-5,7,9,9,-4,-8,-34,-30,-9,-4,5,4,10,-14,-31,-21,-6,-9,-2,7,6,-6,-49,-13,-14,-6,-1,15,-10,-32,-3,-6,-9,7,12,14,-13,4],
# [-10,3,4,9,16,4,3,2,-12,-16,9,18,9,8,6,16,-23,-7,1,19,24,32,25,31,-12,-8,3,10,21,23,16,16,-13,-9,-9,4,11,12,13,9,-4,-2,-3,-4,4,8,18,2,-16,-16,2,3,2,2,-22,-14,-12,-15,-19,1,-11,-27,-7,-11],
# [-4,-4,0,-4,0,0,5,4,0,6,4,4,1,4,2,4,4,8,9,5,4,3,8,1,2,8,7,10,5,6,6,-6,-1,4,0,3,-1,1,0,-8,-2,-2,-3,-6,-20,-2,-2,-9,-4,-3,3,-19,-39,-30,16,15,-18,50,6,-22,-7,-56,33,20],
# [0,0,0,0,0,0,0,0,91,105,94,87,68,86,75,92,34,57,44,29,32,19,20,20,-4,-7,-23,-34,-40,-29,-6,-23,-18,-34,-34,-45,-45,-39,-24,-33,-30,-29,-41,-33,-36,-35,-33,-36,-20,-20,-27,-28,-19,-34,-25,-35,0,0,0,0,0,0,0,0],
# [-8,-5,-5,0,0,-5,-5,-12,-13,-2,8,5,9,-5,-11,-10,-15,2,26,20,15,22,3,-6,-7,5,22,31,0,18,7,-7,-7,-13,21,19,15,9,-5,-5,-11,-11,0,10,5,-4,-5,-13,-9,-16,-21,-9,-1,3,-1,-9,-13,-16,-20,-22,-10,-15,-11,-5],
# [-2,-6,0,-8,3,-1,-1,-1,-6,4,0,-1,0,0,3,3,0,2,-2,0,0,20,10,-2,-4,9,8,11,8,7,-3,8,-3,2,16,8,3,2,-8,-10,-2,-7,14,20,16,-2,-11,-3,-10,-5,-1,5,-14,-4,-20,-15,-11,-8,-24,-20,-8,-19,-6,-5],
# [6,15,11,12,11,12,8,10,15,19,15,18,18,16,13,14,14,15,14,12,12,17,13,2,11,11,15,11,10,11,2,4,3,4,3,3,-3,-1,-10,-24,2,-13,-14,-14,-15,-11,-16,-25,-25,-12,-15,-14,-15,-10,-24,-27,-18,-16,-10,-17,-21,-22,-25,-38],
# [-21,-4,2,13,13,-4,-11,-9,-6,-4,9,15,13,4,4,2,-14,1,6,19,22,16,7,10,-5,2,5,13,22,21,11,11,-7,-3,0,17,1,8,2,0,-4,-7,-7,-7,-7,-8,0,-3,-7,-14,-14,-24,-15,-16,-18,-12,-10,-19,-22,-35,-23,-32,-23,-33],
# [-34,-28,-8,-21,-12,5,19,16,-7,14,15,7,5,24,14,10,4,6,23,5,6,8,21,0,2,16,22,17,13,24,7,-12,-10,0,21,11,19,18,0,-21,-11,-6,10,6,21,13,-3,-16,-18,-12,8,7,11,12,-7,-21,-37,-41,-16,-20,-19,-4,-32,-40],

[0,0,0,0,0,0,0,0,43,42,41,46,29,24,24,30,17,11,17,28,29,19,19,-3,-32,-16,-16,-6,-3,-14,-12,-15,-43,-31,-24,-10,-15,-15,-12,-33,-43,-31,-35,-29,-26,-18,0,-31,-47,-39,-36,-68,-57,-1,8,-35,0,0,0,0,0,0,0,0],
[-8,-4,-1,0,3,0,0,-8,-16,-9,8,7,7,8,-4,-11,-8,8,18,15,19,27,10,6,-3,11,9,34,34,58,6,0,-24,1,18,11,30,23,10,-9,-35,-16,0,5,11,6,16,-22,-6,-22,-16,-4,-7,4,-4,-5,-5,-37,-14,-34,-14,-12,-27,-1],
[0,0,0,-4,0,0,0,0,-16,0,-10,-4,0,5,0,-7,-5,8,-1,5,6,15,11,20,-6,1,3,28,16,19,3,-4,-8,0,10,34,23,8,-4,-9,-1,3,8,0,0,-2,4,3,1,-6,0,-10,0,10,11,-15,-13,-12,-19,-19,-16,-26,-6,-3],
[3,2,7,12,3,5,10,10,-6,-4,21,24,20,18,8,10,-15,-4,6,11,13,17,16,5,-18,-7,-5,7,9,9,-4,-8,-34,-30,-9,-4,5,4,10,-14,-31,-21,-6,-9,-2,7,6,-6,-49,-13,-14,-6,-1,15,-10,-32,-3,-6,-9,7,13,14,-13,4],
[-10,3,4,9,16,4,3,2,-12,-16,9,18,9,8,6,16,-23,-7,1,19,24,32,25,31,-12,-8,3,10,21,23,16,16,-13,-9,-9,4,11,12,13,9,-4,-2,-3,-4,4,8,18,2,-16,-16,2,3,2,2,-22,-14,-12,-15,-19,1,-11,-27,-7,-11],
[-4,-4,0,-4,0,0,5,4,0,6,4,4,1,4,2,4,4,8,9,5,4,3,8,1,2,8,7,10,5,6,6,-6,-1,4,0,3,-1,1,0,-8,-2,-2,-3,-6,-20,-2,-2,-9,-4,-3,3,-19,-39,-30,16,15,-18,50,6,-22,-7,-56,33,20],
[0,0,0,0,0,0,0,0,90,104,93,86,68,85,74,91,34,57,44,29,32,19,20,20,-4,-7,-23,-34,-40,-29,-6,-23,-18,-34,-34,-45,-45,-39,-24,-33,-30,-29,-41,-33,-36,-35,-33,-36,-20,-20,-27,-28,-19,-34,-25,-35,0,0,0,0,0,0,0,0],
[-8,-5,-5,0,0,-5,-5,-12,-13,-2,8,5,9,-5,-11,-10,-15,2,26,20,15,22,3,-6,-7,5,22,31,0,18,7,-7,-7,-13,21,19,15,9,-5,-5,-11,-11,0,10,5,-4,-5,-13,-9,-16,-21,-9,-1,3,-1,-9,-13,-16,-20,-22,-10,-15,-11,-5],
[-2,-6,0,-8,3,-1,-1,-1,-6,4,0,-1,0,0,3,3,0,2,-2,0,0,20,10,-2,-4,9,8,11,8,7,-3,8,-3,2,16,8,3,2,-8,-10,-2,-7,14,20,16,-2,-11,-3,-10,-5,-1,4,-14,-4,-20,-15,-11,-8,-24,-20,-8,-19,-6,-5],
[6,15,11,12,11,12,8,10,15,19,15,18,18,16,13,14,14,15,14,12,12,17,13,2,11,11,15,11,10,11,2,4,3,4,3,3,-3,-1,-10,-24,2,-13,-14,-14,-15,-11,-16,-25,-25,-12,-15,-14,-15,-10,-24,-27,-18,-16,-10,-17,-21,-22,-25,-38],
[-21,-4,2,13,13,-4,-11,-9,-6,-4,9,15,13,4,4,2,-14,1,6,19,22,16,7,10,-5,2,5,13,22,21,11,11,-7,-3,0,17,1,8,2,0,-4,-7,-7,-7,-7,-8,0,-3,-7,-14,-14,-24,-15,-16,-18,-12,-10,-19,-22,-35,-23,-32,-23,-33],
[-34,-28,-8,-21,-12,5,19,16,-7,14,15,7,5,24,14,10,4,6,23,5,6,8,21,0,2,16,22,17,13,24,7,-12,-10,0,21,11,19,18,0,-21,-11,-6,10,6,21,13,-3,-16,-18,-12,8,7,11,12,-7,-21,-37,-41,-16,-20,-19,-4,-32,-40],







# 
#     # MIDGAME pawn
#     [ 
#         0, 0, 0, 0, 0, 0, 0, 0,
#         43, 42, 41, 46, 29, 24, 24, 30,
#         17, 12, 17, 28, 29, 19, 19, -3,
#         -32, -16, -16, -6, -3, -14, -12, -15,
#         -43, -31, -24, -10, -15, -16, -12, -33,
#         -43, -32, -35, -30, -27, -19, -1, -31,
#         -47, -39, -36, -68, -57, -1, 9, -35,
#         0, 0, 0, 0, 0, 0, 0, 0,
#     ],
#     # MIDGAME knight
#     [ 
#         -8, -4, -1, 0, 3, 0, 0, -8,
#         -16, -9, 8, 7, 7, 8, -4, -11,
#         -8, 8, 18, 15, 19, 27, 10, 6,
#         -3, 11, 9, 34, 34, 58, 6, 0,
#         -24, 1, 18, 11, 30, 23, 10, -10,
#         -35, -16, 1, 5, 11, 6, 16, -22,
#         -6, -22, -16, -5, -8, 4, -4, -5,
#         -5, -37, -14, -34, -14, -12, -27, -1,
#     ],
#     # MIDGAME bishop
#     [  
#         0, 0, 0, -4, 0, 0, 0, 0,
#         -16, 0, -10, -4, 0, 5, 0, -7,
#         -5, 8, -1, 5, 6, 15, 11, 20,
#         -6, 2, 3, 28, 16, 19, 4, -4,
#         -8, 1, 11, 34, 23, 9, -4, -9,
#         -1, 4, 9, 1, 0, -2, 5, 4,
#         2, -6, 1, -10, 0, 10, 11, -15,
#         -13, -12, -19, -19, -16, -26, -6, -3,
#     ],
#     # MIDGAME rook
#     [  
#         3, 2, 7, 12, 3, 5, 10, 10,
#         -6, -4, 21, 24, 20, 18, 8, 10,
#         -15, -4, 6, 11, 13, 17, 16, 5,
#         -18, -7, -5, 7, 9, 9, -4, -8,
#         -34, -30, -9, -4, 5, 4, 10, -14,
#         -31, -21, -6, -9, -2, 7, 6, -6,
#         -49, -13, -14, -6, -2, 15, -10, -32,
#         -4, -6, -10, 8, 13, 15, -13, 4,
#     ],
#     # MIDGAME queen
#     [
#         -10, 3, 4, 9, 16, 4, 3, 2,
#         -12, -16, 9, 18, 9, 8, 6, 16,
#         -23, -7, 1, 19, 24, 32, 25, 31,
#         -12, -8, 3, 10, 21, 23, 16, 16,
#         -13, -9, -9, 4, 11, 12, 13, 9,
#         -4, -3, -3, -4, 4, 9, 18, 2,
#         -16, -16, 3, 3, 2, 2, -22, -14,
#         -12, -15, -19, 1, -11, -27, -7, -11,
#     ],
#     # MIDGAME king
#     [  
#         -4, -4, 0, -4, 0, 0, 5, 4,
#         0, 6, 4, 4, 1, 4, 2, 4,
#         4, 8, 9, 5, 4, 3, 8, 1,
#         2, 8, 7, 10, 5, 6, 6, -6,
#         -1, 4, 0, 3, -1, 1, 0, -8,
#         -2, -2, -3, -6, -20, -2, -3, -9,
#         -4, -3, 3, -19, -39, -30, 16, 15,
#         -18, 50, 6, -22, -8, -56, 34, 20,
#     ],
# 
#     # ENDGAME pawn
#     [  
#         0,0,0,0,0,0,0,0,
#         90, 104, 93, 86, 68, 85, 74, 91,
#         34, 57, 44, 29, 32, 19, 20, 20,
#         -4, -7, -23, -34, -40, -29, -6, -23,
#         -18, -34, -34, -45, -45, -39, -24, -33,
#         -31, -30, -41, -33, -36, -35, -34, -36,
#         -20, -20, -27, -28, -19, -34, -25, -35,
#         0,0,0,0,0,0,0,0,
#     ],
#     # ENDGAME knight
#     [
#         -8, -5, -5,0,0, -5, -5, -12,
#         -13, -2,8,5,9, -5, -11, -10,
#         -15,2, 26, 20, 15, 22,3, -6,
#         -7,5, 22, 31,0, 18,7, -7,
#         -7, -13, 21, 19, 15,9, -5, -5,
#         -11, -11,0, 10,5, -5, -5, -13,
#         -9, -16, -21, -10, -2,3, -1, -9,
#         -13, -16, -20, -22, -10, -15, -11, -5,
#     ],
#     # ENDGAME bishop
#     [  
#         -2, -6,0, -8,3, -1, -1, -1,
#         -6,4,0, -1,1,0,4,4,
#         1,3, -2,1,1, 21, 10, -2,
#         -4,9,8, 11,8,8, -3,8,
#         -3,3, 17,9,4,3, -8, -10,
#         -2, -7, 15, 21, 17, -2, -11, -3,
#         -10, -5, -1,5, -14, -4, -20, -15,
#         -11, -8, -24, -20, -8, -19, -6, -5,
#     ],
#     # ENDGAME rook
#     [  
#         6, 15, 11, 12, 11, 12,8, 10,
#         15, 19, 15, 18, 18, 16, 13, 14,
#         14, 15, 14, 12, 12, 17, 13,2,
#         11, 11, 15, 11, 10, 11,2,4,
#         3,4,3,4, -3, -1, -10, -24,
#         2, -13, -14, -14, -15, -11, -16, -25,
#         -25, -12, -15, -14, -15, -10, -24, -27,
#         -18, -16, -11, -17, -21, -22, -25, -38,
#     ],
#     # ENDGAME queen
#     [ 
#         -21, -4,2, 13, 13, -4, -11, -9,
#         -6, -4,9, 15, 13,4,4,2,
#         -14,1,6, 19, 22, 16,7, 10,
#         -5,2,5, 13, 22, 21, 11, 11,
#         -7, -3,0, 17,2,8,2,0,
#         -4, -7, -7, -7, -7, -8,0, -3,
#         -7, -14, -14, -24, -15, -16, -18, -12,
#         -10, -19, -22, -35, -23, -32, -23, -33,
#     ],
#     # ENDGAME king
#     [  
#         -34, -28, -8, -21, -12,5, 19, 16,
#         -7, 14, 15,7,5, 24, 14, 10,
#         4,6, 23,5,6,8, 21, -1,
#         2, 16, 22, 17, 13, 24,7, -12,
#         -10,0, 21, 11, 19, 18,0, -21,
#         -11, -6, 10,6, 21, 13, -4, -16,
#         -18, -12,8,7, 11, 13, -8, -21,
#         -37, -41, -16, -20, -19, -5, -32, -40,
#     ],
    # Pawn shield MG
    [
#         -30, 0, 5, -10, -15, 0, 0, 0,
#         -30, 0, -5, -15, -15, 0, 0, 0,
#         -30, 0, -5, -10, -15, 0, 0, 0,
#         0, 0, 0, 0, 0, 0, 0, 0,
#         0, 0, 0, 0, 0, 0, 0, 0,
#         -30, 0, -5, -10, -15, 0, 0, 0,
#         -30, 0, -5, -15, -15, 0, 0, 0,
#         -30, 0, 5, -10, -15, 0, 0, 0,

#         15, 0, 20, 2, 0, 0, 0, 0,
#         -16, 0, -14, -35, 0, 0, 0, 0, 
#         -14, 0, -13, -15, 0, 0, 0, 0, 
#         0, 0, 0, 0, 0, 0, 0, 0, 
#         0, 0, 0, 0, 0, 0, 0, 0,
#         -1, 0, 1, 3, 0, 0, 0, 0, 
#         -33, 0, -11, -34, 0, 0, 0, 0, 
#         11, 0, 20, -10, 0, 0, 0, 0, 

        15, 0, 20, 2, 0, 0, 0, 0,
        -25, 0, -14, -20, 0, 0, 0, 0, 
        -14, 0, -13, -15, 0, 0, 0, 0, 
        0, 0, 0, 0, 0, 0, 0, 0, 
        0, 0, 0, 0, 0, 0, 0, 0,
        -14, 0, 1, 3, 0, 0, 0, 0, 
        -33, 0, -11, -20, 0, 0, 0, 0, 
        11, 0, 20, -10, 0, 0, 0, 0, 
    ],
]

# Rescales the values in a table to fit within a given range. The range is compress equally from both ends.
def rescale_table(table: [int], range_limit: int) -> [int]:
    min_val = min(table)
    max_val = max(table)
    diff = max_val - min_val
    rescale_range = diff
    
    if rescale_range > range_limit:
        compress_amount = (rescale_range - range_limit) // 2
        rescale_range = range_limit
        
        new_min = min_val + compress_amount
        table = [int((x-min_val)/diff * rescale_range + new_min) for x in table]
    return table


def generate_bitmaps(table: [int], range_limit: int) -> [int]:
    min_val = min(table)
    max_val = max(table)
    diff = max_val - min_val
    
    table = [x-min_val for x in table]
    print(min_val, max_val, diff, table[:8])
    
    bitmaps = [0] * 6
    for squareIdx, value in enumerate(table):
        for i in range(5,-1,-1):
            if value >= 2**i:
                value -= 2**i
                bitmaps[i] |= 1 << squareIdx
                
#     for i, bitmap in enumerate(bitmaps):
#         print(f"  {2**i}\t{bitmap:064b} {bitmap}")
    
    return (bitmaps, min_val)
    
    
def main():
    absolute_min = min([min(table) for table in tables])
#     absolute_max = max([max(table) for table in tables])
#     absolute_diff = absolute_max - absolute_min
#     print(absolute_min, absolute_max, absolute_diff)
#     return

    i = 0
    final_bitmaps = []
    final_min_vals = []
    for table in tables:
        piece = pieces[i%6]
#         phase = phases[i//6]
        print(piece)
        
        rescaled_table = rescale_table(table, 64)
        if rescaled_table != table:
            print(f"WARNING: table {i} rescaled from {max(table)-min(table)} to {max(rescaled_table)-min(rescaled_table)}")
            
        (bitmaps, min_val) = generate_bitmaps(rescaled_table, 64)
        final_bitmaps.extend(bitmaps)
        final_min_vals.append(min_val)
        i+=1

    final_min_vals = [x-absolute_min for x in final_min_vals] # Account for abs min being hardcoded
    final_min_vals = [0] + final_min_vals # prepend absolute_min (which is normalized to 0)
    final_min_vals += [0, 0, 1, 1, 2, 4, 0] # piece phase values
    final_min_vals += [0, 0, 4, 7, 7, 14, 0] # piece maximum mobility half values
    final_min_vals += [0, 0, 5, 3, 3, 0, 0] # piece mobility scores (mg)
    final_min_vals += [0, 0, 2, 3, 2, 6, 0] # piece mobility scores (mg)
    final_min_vals += [x+42 for x in [0, 3, -16, -13, 2, -39, 0]] # material endgame values
    print('final_min_vals', [x+0 for x in final_min_vals], '(includes baseline table min value)') 
    
    (bitmaps, min_val) = generate_bitmaps(final_min_vals, 64)
    print('ABSOLUTE MIN:', absolute_min) 
    final_bitmaps = bitmaps + final_bitmaps
#     final_min_vals = [min_val] + final_min_vals 
    
    print('final_min_vals', [x+0 for x in final_min_vals], '(includes baseline table min value)') 

    for i, bitmap in enumerate(bitmaps):
        print(f"  {2**i}\t{bitmap:064b} {bitmap}")
    
    dynamic_data = [0] * (1+len(tables))*64
#     dynamic_data[0] = -68
    for bitmapIdx in range((1+len(tables))*6):
        if bitmapIdx % 6 == 0 and bitmapIdx // 6 != 0:
            offset = dynamic_data[bitmapIdx//6] # Only want to exec once for each 0..6 bitmap chunk
            for i in range(64):
                dynamic_data[bitmapIdx // 6 * 64 + i] = offset -68
        for i in range(64):
            dynamic_data[bitmapIdx // 6 * 64 + i] += 1 << (bitmapIdx % 6) if final_bitmaps[bitmapIdx] & (1 << i) else 0
        
    j=0
    for idx in range(0, len(dynamic_data), 8):
        print(dynamic_data[idx:idx+8])
        j += 1
        if j % 8 == 0:
            print()
            
    names = ['baselines + constants']
    names += [f'{phase} {piece}' for phase in phases for piece in pieces]
    names += ['pawn shield mg']
    for i, bitmap in enumerate(final_bitmaps):
        print(f'0x{bitmap:016x}, ', end='')
        if i % 6 == 5:
            print('//', names[i//6])
    print('Absolute Min:   ', absolute_min)
    print('True Zero Offset:', -63-absolute_min)
    
    mat_mg = [0, 100, 332, 343, 433, 1030, 0]
    mat_eg = [0, 125, 210, 243, 454, 725, 0]
    orig_mat_diff = [x-y for x,y in zip(mat_eg, mat_mg)]
    print('orig\t', orig_mat_diff)
    F = 8
    mat_diff = [max(absolute_min, x//F) for x in orig_mat_diff]
    print('m dif\t', mat_diff)
    mat_diff_unpack = [F * x for x in mat_diff]
    print('unpk\t', mat_diff_unpack)
    orig_unpack_diff = [abs(x-y) for x,y in zip(mat_diff_unpack, orig_mat_diff)]
    print('err\t', orig_unpack_diff)
    
    
    mat_mg = [100, 332, 343, 433, 1030, 0]
    mat_eg = [125, 210, 243, 454, 725, 0]
    print('Current weights:')
    
    for x in mat_mg: print(x, end=',')
    for table in tables[0:6]: 
        for x in table:
            print(x, end=',')
    print(24, end=',') # bishop pair mg
    
    for x in mat_eg: print(x, end=',')
    for table in tables[6:12]: 
        for x in table:
            print(x, end=',')
    print(41) # bishop pair eg
    
    
    # LMR table test
#     print('d/m:\t', end='')
#     for numMoves in range(3, 50):
#         print(f'{numMoves:2} ', end=' ')
#     print()
#     for depth in range(3, 30):
#         print(f'{depth}:\t', end='')
#         for numMoves in range(3, 50):
#             value = max(2, depth//4) + numMoves//12
#             print(f'{value:02} ', end=' ')
#         print()
    
    # depth * 30 + moveIndex
#     MOVE_COUNT_LIMIT = 30
#     DEPTH_LIMIT = 30
#     lmr = [0] * (MOVE_COUNT_LIMIT * DEPTH_LIMIT)
#     for i in range(MOVE_COUNT_LIMIT * DEPTH_LIMIT):
#         depth = i // MOVE_COUNT_LIMIT
#         moveIndex = i % MOVE_COUNT_LIMIT
#         value = max(2, depth//4) + moveIndex//12
#         lmr[i] = value
#     print('LMR table:')
#     for i in range(0, len(lmr), MOVE_COUNT_LIMIT):
#         print(lmr[i:i+MOVE_COUNT_LIMIT])
   
    '''
    int i = 0
    while(i < 1000) lmr[i++] = Math.max(2, i/64/4) + i&63/12; // 16
    '''
    
#     depth = 12, moveCount = 6 
#     lmrIndex = depth * 30 + moveCount

if __name__ == "__main__":
    main()